File: instantsearch.voice/src/main/java/com/algolia/instantsearch/voice/ui/PermissionDialogFragment.java
package com.algolia.instantsearch.voice.ui;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentActivity;

import static android.Manifest.permission.RECORD_AUDIO;

public class PermissionDialogFragment extends DialogFragment {
    public static final int ID_REQ_VOICE_PERM = 1;
    public static final String DEFAULT_TITLE = "You can use voice search to find results";
    public static final String DEFAULT_MESSAGE = "Can we access your device's microphone to enable voice search?";
    public static final String DEFAULT_POSITIVE_BUTTON = "Allow microphone access";
    public static final String DEFAULT_NEGATIVE_BUTTON = "No";

    private String title = DEFAULT_TITLE;
    private String message = DEFAULT_MESSAGE;
    private String positiveButton = DEFAULT_POSITIVE_BUTTON;
    private String negativeButton = DEFAULT_NEGATIVE_BUTTON;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final FragmentActivity activity = getActivity();
        if (activity == null) {
            throw new IllegalStateException("PermissionDialogFragment must be used within an activity.");
        }
        return new AlertDialog.Builder(activity)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(positiveButton, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        ActivityCompat.requestPermissions(activity, new String[]{RECORD_AUDIO}, ID_REQ_VOICE_PERM);
                        dismiss();
                    }
                })
                .setNegativeButton(negativeButton, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                })
                .create();
    }

    //region setters
    public void setTitle(String title) {
        this.title = title;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public void setPositiveButton(String positiveButton) {
        this.positiveButton = positiveButton;
    }

    public void setNegativeButton(String negativeButton) {
        this.negativeButton = negativeButton;
    }
    //endregion
}

File: instantsearch.voice/src/main/java/com/algolia/instantsearch/voice/ui/PermissionDialogFragment.kt
package com.algolia.instantsearch.voice.ui

import android.Manifest.permission.RECORD_AUDIO
import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.app.DialogFragment

class PermissionDialogFragment : DialogFragment() {

    private enum class Argument {
        Title,
        Message,
        PositiveButton,
        NegativeButton
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments == null) {
            arguments = buildArguments()
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val bundle = arguments!!
        val title = bundle.getString(Argument.Title.name)
        val message = bundle.getString(Argument.Message.name)
        val positiveButton = bundle.getString(Argument.PositiveButton.name)
        val negativeButton = bundle.getString(Argument.NegativeButton.name)

        return AlertDialog.Builder(requireContext())
            .setTitle(title)
            .setMessage(message)
            .setPositiveButton(positiveButton) { _, _ ->
                ActivityCompat.requestPermissions(requireActivity(), arrayOf(RECORD_AUDIO), PermissionRequestRecordAudio)
                dismiss()
            }
            .setNegativeButton(negativeButton) { dialog, _ -> dialog.cancel() }
            .create()
    }

    companion object {

        const val PermissionRequestRecordAudio = 1

        @JvmOverloads
        fun buildArguments(
            title: String = "You can use voice search to find results",
            message: String = "Can we access your device's microphone to enable voice search?",
            positiveButton: String = "Allow microphone access",
            negativeButton: String = "No"
        ) = Bundle().also {
            it.putString(Argument.Title.name, title)
            it.putString(Argument.Message.name, message)
            it.putString(Argument.PositiveButton.name, positiveButton)
            it.putString(Argument.NegativeButton.name, negativeButton)
        }
    }
}
